### PR TITLE
[5.1] Add query builder resolver

### PR DESCRIPTION
**Pull request by vlakoff**
People may want to extend `Query\Builder` in order to add methods, etc.
Currently, although not _impossible_, it is incredibly heavy and hacky to do so, as the class is hardcoded.

The attached commit is simply a POC. Usage example:

```
DB::setQueryBuilderResolver(function ($connection, $grammar, $processor) {
    return new CustomQueryBuilder($connection, $grammar, $processor);
});
```

On a related note, `$grammar` and `$processor` parameters of  `Query\Builder::__construct()` may be made optional, as we can get them from `$connection`. This would let simplifying the above code.

Personally, I'd even add the possibility to pass a class name:

```
DB::setQueryBuilderResolver(CustomQueryBuilder::class);
```

Additional notes:
- Inspired by, for instance, 7584d0f813edd9762d7558e218be3cebdb502d09.
- Similar work may be done for `Schema\Builder`, mainly for consistency.

### CODE REVIEW

File: src/Illuminate/Database/Connection.php [hunk 1]
```diff
@@ -134,6 +134,13 @@ class Connection implements ConnectionInterface
     protected $config = [];
 
     /**
+     * The Query Builder resolver callback.
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Connection.php]: query builder

File: src/Illuminate/Database/Connection.php [hunk 2]
```diff
@@ -938,6 +950,17 @@ public function getQueryGrammar()
     }
 
     /**
+     * Set the Query Builder resolver callback.
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Connection.php]: lower case for nouns please

**Comment by JosephSilber** [comment]
Why not just make the query builder [macroable](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Support/Traits/Macroable.php)?

**Comment by JosephSilber** [comment]
Hacky? How's it any hackier than any other class using the `Macroable` trait?

**Comment by GrahamCampbell** [comment]
Like, take for example the string class. That's stateless by design, so I don't think the macro stuff there is hacky.

**Comment by vlakoff** [comment]
Macroable seems fine too. Rather, wonderful. Both techniques seem acceptable to me. Your pick.

Do you agree on the following points?

* Many people may wish to extend the query builder, so anyway, we should provide a convenient way to do so (as I said, currently it's very heavy and dirty in order to extend it).
* Having to pass grammar and processor to [`Query\Builder::__construct()`](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Database/Query/Builder.php) is redundant most (if not all) of the time. We could make these parameters optional. For instance, [`Schema\Builder::__construct()`](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Database/Schema/Builder.php) infers grammar from the connection.

**Comment by taylorotwell** [comment]
The two things aren't even really mutually exclusive. However, the query builder already has a __call method, I'm not sure how to make it jive with Macroable... I assume you can override it and combine them?

**Comment by vlakoff** [comment]
I don't know why I hadn't spotted it in my tests, but it seems that, unlike regular methods, the class __call() has precedence over the trait __call():

    trait Sometrait {
        public function __call($method, $params) { return 'Sometrait::__call()'; }
    }
    class Someclass {
        use Sometrait;
        public function __call($method, $params) { return 'Someclass::__call()'; }
    }
    (new Someclass)->blahblah(); // 'Someclass::__call()'

As a result, the query builder triggers its own __call() and the macro never gets triggered…

I'm unaware of a way to call the trait __call() from inside the class__call(), is it possible?

**Comment by JosephSilber** [comment]
@taylorotwell https://github.com/laravel/framework/blob/5.1/src/Illuminate/Cache/Repository.php#L16-L18

**Comment by vlakoff** [comment]
Looks like this is possible, see [this answer](http://stackoverflow.com/questions/11939166/how-to-override-trait-function-and-call-it-from-the-overriden-function#11939306). (thanks ircmaxell!)

I have implemented it successfully. Good thing, if the method doesn't exist we get a `BadMethodCallException` too, only with a different message (the one from `Macroable` instead of the one from `Builder`, who becomes unreachable).

Though, I haven't examined the situation about `where*()` calls.

Another possibility would be to lookup `$this->macros` from inside the `Builder` class.

**Comment by JosephSilber** [comment]
@vlakoff you're overthinking this a little. After aliasing `Macroable@__call`, all you need is to add these 3 lines the the beginning of `Builder`'s `__call` method:

```php
public function __call($method, $parameters)
{
    if (static::hasMacro($method)) {
        return $this->macroCall($method, $parameters);
    }
    
    if (Str::startsWith($method, 'where')) {
        return $this->dynamicWhere($method, $parameters);
    }

    $className = get_class($this);

    throw new BadMethodCallException("Call to undefined method {$className}::{$method}()");
}
```

**Comment by vlakoff** [comment]
Indeed. This solution looks really great. :+1:

**Comment by taylorotwell** [comment]
Query macros have been added to 5.1.

**Comment by vlakoff** [comment]
For reference, implemented in https://github.com/laravel/framework/commit/60ecc104741e083e37d01fcde824abf853537a5a.

**Pull request closed without merging**